File: packages/b2c-tooling-sdk/src/auth/oauth.ts
/*
 * Copyright (c) 2025, Salesforce, Inc.
 * SPDX-License-Identifier: Apache-2
 * For full license text, see the license.txt file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
 */
import type {AuthStrategy, AccessTokenResponse, DecodedJWT, FetchInit} from './types.js';
import {dispatchFetch} from './dispatch-fetch.js';
import {wrapNetworkError} from '../errors/network-error.js';
import {getLogger} from '../logging/logger.js';
import {DEFAULT_ACCOUNT_MANAGER_HOST} from '../defaults.js';
import {globalAuthMiddlewareRegistry, applyAuthRequestMiddleware, applyAuthResponseMiddleware} from './middleware.js';

// Module-level token cache to support multiple instances with same clientId
const ACCESS_TOKEN_CACHE: Map<string, AccessTokenResponse> = new Map();

// In-flight token requests, keyed by cacheKey, so concurrent callers coalesce
// onto a single token-endpoint round trip instead of stampeding the server.
const PENDING_TOKEN_REQUESTS: Map<string, Promise<AccessTokenResponse>> = new Map();

export interface OAuthConfig {
  clientId: string;
  clientSecret: string;
  scopes?: string[];
  accountManagerHost?: string;
}

/**
 * Decodes a JWT token without verification.
 * Exported for use by other auth strategies.
 */
export function decodeJWT(jwt: string): DecodedJWT {
  const parts = jwt.split('.');
  if (parts.length !== 3) {
    throw new Error('Invalid JWT format');
  }
  const header = JSON.parse(Buffer.from(parts[0], 'base64').toString());
  const payload = JSON.parse(Buffer.from(parts[1], 'base64').toString());
  return {header, payload};
}

/**
 * Generates a cache key for OAuth tokens.
 * Includes auth method to distinguish between client-credentials and JWT tokens.
 *
 * @param clientId - OAuth client ID
 * @param method - Authentication method (client-credentials or jwt)
 * @param accountManagerHost - Account Manager hostname
 * @param scopes - OAuth scopes (optional)
 * @returns Cache key string
 */
export function getOAuthCacheKey(
  clientId: string,
  method: 'client-credentials' | 'jwt',
  accountManagerHost: string,
  scopes?: string[],
): string {
  const scopesKey = scopes?.sort().join(',') || '';
  return `${accountManagerHost}:${clientId}:${method}:${scopesKey}`;
}

/**
 * Gets a cached OAuth token if valid.
 *
 * @param cacheKey - Cache key from getOAuthCacheKey()
 * @param requiredScopes - Scopes that must be present in the cached token
 * @returns Cached token response if valid, undefined otherwise
 */
export function getCachedOAuthToken(cacheKey: string, requiredScopes: string[] = []): AccessTokenResponse | undefined {
  const cached = ACCESS_TOKEN_CACHE.get(cacheKey);
  if (!cached) return undefined;

  const now = new Date();
  const hasAllScopes = requiredScopes.every((scope) => cached.scopes.includes(scope));

  // Check if token is expired or missing required scopes
  if (!hasAllScopes || now.getTime() > cached.expires.getTime()) {
    ACCESS_TOKEN_CACHE.delete(cacheKey);
    return undefined;
  }

  return cached;
}

/**
 * Stores an OAuth token in the global cache.
 *
 * @param cacheKey - Cache key from getOAuthCacheKey()
 * @param tokenResponse - Token response to cache
 */
export function setCachedOAuthToken(cacheKey: string, tokenResponse: AccessTokenResponse): void {
  ACCESS_TOKEN_CACHE.set(cacheKey, tokenResponse);
}

/**
 * Invalidates a cached OAuth token.
 *
 * @param cacheKey - Cache key from getOAuthCacheKey()
 */
export function invalidateCachedOAuthToken(cacheKey: string): void {
  ACCESS_TOKEN_CACHE.delete(cacheKey);
}

/**
 * OAuth 2.0 Client Credentials authentication strategy.
 *
 * Implements the client credentials flow for automated/server-side authentication
 * with no user interaction required. Automatically manages token caching, expiration,
 * and 401 retry logic with single-flight token requests to prevent thundering herd
 * on the token endpoint.
 *
 * @example
 * ```typescript
 * import { OAuthStrategy } from '@salesforce/b2c-tooling-sdk';
 *
 * const auth = new OAuthStrategy({
 *   clientId: 'your-client-id',
 *   clientSecret: 'your-client-secret',
 *   scopes: ['sfcc.products'],
 * });
 *
 * const response = await auth.fetch('https://api.example.com/products');
 * ```
 */
export class OAuthStrategy implements AuthStrategy {
  private accountManagerHost: string;
  private _hasHadSuccess = false;
  private cacheKey: string;

  /**
   * Creates a new OAuthStrategy instance with the provided OAuth configuration.
   *
   * @param config - OAuth client credentials and optional configuration
   */
  constructor(private config: OAuthConfig) {
    this.accountManagerHost = config.accountManagerHost || DEFAULT_ACCOUNT_MANAGER_HOST;
    this.cacheKey = getOAuthCacheKey(
      this.config.clientId,
      'client-credentials',
      this.accountManagerHost,
      this.config.scopes,
    );
  }

  /**
   * Performs a fetch request with OAuth bearer token authentication.
   *
   * Automatically injects the Authorization header and client ID header with a valid
   * access token. Implements 401 retry logic: if a previously-successful request
   * returns 401, invalidates the cached token and retries once with a fresh token.
   * Does not retry on initial 401 to avoid retrying with bad credentials.
   *
   * @param url - The URL to fetch
   * @param init - Optional fetch init options (headers, body, method, etc.)
   * @returns The fetch response
   */
  async fetch(url: string, init: FetchInit = {}): Promise<Response> {
    const token = await this.getAccessToken();

    const headers = new Headers(init.headers);
    headers.set('Authorization', `Bearer ${token}`);
    headers.set('x-dw-client-id', this.config.clientId);

    // Pass through dispatcher for TLS/mTLS support (see dispatchFetch)
    let res = await dispatchFetch(url, {...init, headers});

    if (res.status !== 401) {
      this._hasHadSuccess = true;
    }

    // RESILIENCE: If we previously had a successful response and now get a 401,
    // the token likely expired. Retry once after invalidating the cached token.
    // Skip retry on initial 401 to avoid retrying with bad credentials.
    if (res.status === 401 && this._hasHadSuccess) {
      this.invalidateToken();
      const newToken = await this.getAccessToken();
      headers.set('Authorization', `Bearer ${newToken}`);
      res = await dispatchFetch(url, {...init, headers});
    }

    return res;
  }

  async getAuthorizationHeader(): Promise<string> {
    const token = await this.getAccessToken();
    return `Bearer ${token}`;
  }

  /**
   * Gets the decoded JWT payload
   */
  async getJWT(): Promise<DecodedJWT> {
    const token = await this.getAccessToken();
    return decodeJWT(token);
  }

  /**
   * Gets the full token response including expiration and scopes.
   * Useful for commands that need to display or return token metadata.
   */
  async getTokenResponse(): Promise<AccessTokenResponse> {
    const logger = getLogger();
    const cached = getCachedOAuthToken(this.cacheKey, this.config.scopes || []);

    if (cached) {
      logger.debug('[OAuthStrategy] Reusing cached access token');
      return cached;
    }

    return this.refreshTokenSingleflight();
  }

  /**
   * Invalidates the cached token, forcing re-authentication on next request
   */
  invalidateToken(): void {
    invalidateCachedOAuthToken(this.cacheKey);
  }

  /**
   * Creates a new OAuthStrategy with additional scopes merged in.
   * Used by clients that have specific scope requirements.
   *
   * @param additionalScopes - Scopes to add to this strategy's existing scopes
   * @returns A new OAuthStrategy instance with merged scopes
   */
  withAdditionalScopes(additionalScopes: string[]): OAuthStrategy {
    const mergedScopes = [...new Set([...(this.config.scopes || []), ...additionalScopes])];
    return new OAuthStrategy({
      ...this.config,
      scopes: mergedScopes,
    });
  }

  /**
   * Gets an access token, using cache if valid
   */
  private async getAccessToken(): Promise<string> {
    const logger = getLogger();
    const cached = getCachedOAuthToken(this.cacheKey, this.config.scopes || []);

    if (cached) {
      logger.debug('[OAuthStrategy] Reusing cached access token');
      return cached.accessToken;
    }

    const tokenResponse = await this.refreshTokenSingleflight();
    return tokenResponse.accessToken;
  }

  /**
   * Returns a fresh token, coalescing concurrent callers onto a single in-flight
   * token request keyed by cacheKey. Prevents stampeding the AM token endpoint
   * when many requests trigger refresh at once.
   */
  private refreshTokenSingleflight(): Promise<AccessTokenResponse> {
    const existing = PENDING_TOKEN_REQUESTS.get(this.cacheKey);
    if (existing) {
      getLogger().debug('[OAuthStrategy] Joining in-flight token request');
      return existing;
    }
    const pending = (async () => {
      getLogger().debug('[OAuthStrategy] Requesting new access token');
      const tokenResponse = await this.clientCredentialsGrant();
      setCachedOAuthToken(this.cacheKey, tokenResponse);
      return tokenResponse;
    })().finally(() => {
      PENDING_TOKEN_REQUESTS.delete(this.cacheKey);
    });
    PENDING_TOKEN_REQUESTS.set(this.cacheKey, pending);
    return pending;
  }

  /**
   * Performs client credentials grant flow
   */
  private async clientCredentialsGrant(): Promise<AccessTokenResponse> {
    const logger = getLogger();
    const url = `https://${this.accountManagerHost}/dwsso/oauth2/access_token`;
    const method = 'POST';

    const params = new URLSearchParams({
      grant_type: 'client_credentials',
    });

    if (this.config.scopes && this.config.scopes.length > 0) {
      params.append('scope', this.config.scopes.join(' '));
    }

    const credentials = Buffer.from(`${this.config.clientId}:${this.config.clientSecret}`).toString('base64');

    // Build request object for middleware
    let request = new Request(url, {
      method,
      headers: {
        Authorization: `Basic ${credentials}`,
        'Content-Type': 'application/x-www-form-urlencoded',
      },
      body: params.toString(),
    });

    // Apply auth middleware (e.g., User-Agent)
    const middleware = globalAuthMiddlewareRegistry.getMiddleware();
    request = await applyAuthRequestMiddleware(request, middleware);

    // Convert headers to object for logging
    const requestHeaders: Record<string, string> = {};
    request.headers.forEach((value, key) => {
      requestHeaders[key] = value;
    });

    logger.debug(
      {clientId: this.config.clientId},
      `[Auth] Using OAuth client_credentials grant for client: ${this.config.clientId}`,
    );
    // Debug: Log request start
    logger.debug({method, url}, `[Auth REQ] ${method} ${url}`);

    // Trace: Log request details
    logger.trace({method, url, headers: requestHeaders, body: params.toString()}, `[Auth REQ BODY] ${method} ${url}`);

    const startTime = Date.now();
    let response: Response;
    try {
      response = await fetch(request);
    } catch (err) {
      const host = new URL(url).host;
      throw wrapNetworkError(err, {operation: 'OAuth token request', host});
    }

    // Apply response middleware
    response = await applyAuthResponseMiddleware(request, response, middleware);

    const duration = Date.now() - startTime;

    // Debug: Log response summary
    logger.debug(
      {method, url, status: response.status, duration},
      `[Auth RESP] ${method} ${url} ${response.status} ${duration}ms`,
    );

    // Get response headers
    const responseHeaders: Record<string, string> = {};
    response.headers.forEach((value, key) => {
      responseHeaders[key] = value;
    });

    if (!response.ok) {
      const errorText = await response.text();
      logger.trace({method, url, headers: responseHeaders, body: errorText}, `[Auth RESP BODY] ${method} ${url}`);
      throw new Error(`Failed to get access token: ${response.status} ${response.statusText} - ${errorText}`);
    }

    const data = (await response.json()) as {
      access_token: string;
      expires_in: number;
      scope?: string;
    };

    // Trace: Log response details
    logger.trace({method, url, headers: responseHeaders, body: data}, `[Auth RESP BODY] ${method} ${url}`);

    const jwt = decodeJWT(data.access_token);
    logger.trace({jwt: jwt.payload}, '[Auth] JWT payload');

    const now = new Date();
    const expiration = new Date(now.getTime() + data.expires_in * 1000);
    const scopes = data.scope?.split(' ') ?? [];

    return {
      accessToken: data.access_token,
      expires: expiration,
      scopes,
    };
  }
}

File: python/b2c-tooling-sdk/src/b2c_tooling_sdk/auth/oauth.py
# Copyright (c) 2025, Salesforce, Inc.
# SPDX-License-Identifier: Apache-2.0
# For full license text, see the LICENSE file in the repo root or http://www.apache.org/licenses/LICENSE-2.0
"""OAuth 2.0 client-credentials strategy and the shared module-level token cache.

Mirrors ``src/auth/oauth.ts``. :class:`OAuthStrategy` implements the client
credentials grant for automated/server-side auth, with token caching, expiry
handling, single-flight token requests (so concurrent callers coalesce onto one
token-endpoint round trip), 401 retry, and scope-cascade resolution.

The token cache and in-flight request maps are module-level so multiple strategy
instances sharing a client id reuse the same tokens, exactly as the TS SDK does.
"""

from __future__ import annotations

import asyncio
from datetime import datetime, timedelta, timezone
from typing import Any
from urllib.parse import urlencode

import httpx

from b2c_tooling_sdk.auth.client_credentials import encode_basic_client_credentials
from b2c_tooling_sdk.auth.dispatch_fetch import dispatch_fetch
from b2c_tooling_sdk.auth.jwt_utils import decode_jwt
from b2c_tooling_sdk.auth.middleware import (
    apply_auth_request_middleware,
    apply_auth_response_middleware,
    global_auth_middleware_registry,
)
from b2c_tooling_sdk.auth.types import AccessTokenResponse, DecodedJWT
from b2c_tooling_sdk.defaults import DEFAULT_ACCOUNT_MANAGER_HOST
from b2c_tooling_sdk.errors.network_error import wrap_network_error
from b2c_tooling_sdk.logging import get_logger

# re-exported for parity with the TS module, which defines decodeJWT here.
__all__ = [
    "OAuthConfig",
    "OAuthStrategy",
    "decode_jwt",
    "get_oauth_cache_key",
    "get_cached_oauth_token",
    "set_cached_oauth_token",
    "find_cached_token_satisfying",
    "invalidate_cached_tokens_for_identity",
    "reset_oauth_cache_for_testing",
]

# Module-level token cache to support multiple instances with the same clientId.
_ACCESS_TOKEN_CACHE: dict[str, AccessTokenResponse] = {}

# In-flight token requests keyed by cache key, so concurrent callers coalesce onto
# a single token-endpoint round trip instead of stampeding the server.
_PENDING_TOKEN_REQUESTS: dict[str, asyncio.Future[AccessTokenResponse]] = {}


class OAuthConfig:
    """Configuration for :class:`OAuthStrategy` (client-credentials grant)."""

    def __init__(
        self,
        client_id: str,
        client_secret: str,
        scopes: list[str] | None = None,
        account_manager_host: str | None = None,
    ) -> None:
        self.client_id = client_id
        self.client_secret = client_secret
        self.scopes = scopes
        self.account_manager_host = account_manager_host


def get_oauth_cache_key(
    client_id: str,
    method: str,
    account_manager_host: str,
    scopes: list[str] | None = None,
) -> str:
    """Build a token cache key. Includes the auth method to keep grants distinct."""
    scopes_key = ",".join(sorted(scopes)) if scopes else ""
    return f"{account_manager_host}:{client_id}:{method}:{scopes_key}"


def get_cached_oauth_token(cache_key: str, required_scopes: list[str] | None = None) -> AccessTokenResponse | None:
    """Return a cached token if present, unexpired, and covering ``required_scopes``."""
    required_scopes = required_scopes or []
    cached = _ACCESS_TOKEN_CACHE.get(cache_key)
    if cached is None:
        return None
    now = datetime.now(tz=timezone.utc)
    has_all_scopes = all(scope in cached.scopes for scope in required_scopes)
    if not has_all_scopes or now > cached.expires:
        _ACCESS_TOKEN_CACHE.pop(cache_key, None)
        return None
    return cached


def set_cached_oauth_token(cache_key: str, token_response: AccessTokenResponse) -> None:
    """Store a token in the global cache."""
    _ACCESS_TOKEN_CACHE[cache_key] = token_response


def find_cached_token_satisfying(
    identity_prefix: str,
    required_scopes: list[str],
) -> AccessTokenResponse | None:
    """Return the first non-expired cached token (matching ``identity_prefix``) whose scopes ⊇ ``required_scopes``.

    Used by cascade resolution: a token granted with broader scopes automatically
    satisfies a later request needing a narrower scope, with no extra AM round trip.
    """
    now = datetime.now(tz=timezone.utc)
    for key in list(_ACCESS_TOKEN_CACHE.keys()):
        if not key.startswith(identity_prefix):
            continue
        entry = _ACCESS_TOKEN_CACHE[key]
        if now > entry.expires:
            _ACCESS_TOKEN_CACHE.pop(key, None)
            continue
        if all(s in entry.scopes for s in required_scopes):
            return entry
    return None


def invalidate_cached_tokens_for_identity(identity_prefix: str) -> None:
    """Evict every cached token for an identity prefix (host:clientId:method:).

    Cascade-resolving strategies cache tokens under merged-scope keys, so deleting
    only the base key on a 401 would leave a rejected merged token cached. Clearing
    by identity prefix evicts all of them so the retry re-requests from AM.
    """
    for key in list(_ACCESS_TOKEN_CACHE.keys()):
        if key.startswith(identity_prefix):
            _ACCESS_TOKEN_CACHE.pop(key, None)


def reset_oauth_cache_for_testing() -> None:
    """Clear the module-level token cache and pending-request map (tests only)."""
    _ACCESS_TOKEN_CACHE.clear()
    _PENDING_TOKEN_REQUESTS.clear()


class OAuthStrategy:
    """OAuth 2.0 client-credentials authentication strategy.

    :example:

    .. code-block:: python

        from b2c_tooling_sdk.auth import OAuthStrategy

        auth = OAuthStrategy(OAuthConfig(
            client_id="your-client-id",
            client_secret="your-client-secret",
            scopes=["sfcc.products"],
        ))
        response = await auth.fetch("https://api.example.com/products")
    """

    def __init__(self, config: OAuthConfig) -> None:
        self._config = config
        self._account_manager_host = config.account_manager_host or DEFAULT_ACCOUNT_MANAGER_HOST
        self._has_had_success = False
        self._cache_key = get_oauth_cache_key(
            config.client_id,
            "client-credentials",
            self._account_manager_host,
            config.scopes,
        )
        self._identity_prefix = f"{self._account_manager_host}:{config.client_id}:client-credentials:"

    async def fetch(
        self,
        url: str,
        *,
        method: str = "GET",
        headers: dict[str, str] | None = None,
        content: Any = None,
        dispatcher: httpx.AsyncBaseTransport | None = None,
        **kwargs: Any,
    ) -> httpx.Response:
        """Perform an authenticated request, injecting a bearer token and retrying once on a post-success 401."""
        token = await self._get_access_token()
        request_headers = dict(headers or {})
        request_headers["Authorization"] = f"Bearer {token}"
        request_headers["x-dw-client-id"] = self._config.client_id

        response = await dispatch_fetch(
            url, method=method, headers=request_headers, content=content, dispatcher=dispatcher, **kwargs
        )

        if response.status_code != 401:
            self._has_had_success = True

        # RESILIENCE: if a previously-successful request now 401s, the token likely
        # expired. Retry once with a fresh token. Skip retry on an initial 401 to
        # avoid retrying with bad credentials.
        if response.status_code == 401 and self._has_had_success:
            self.invalidate_token()
            new_token = await self._get_access_token()
            request_headers["Authorization"] = f"Bearer {new_token}"
            response = await dispatch_fetch(
                url, method=method, headers=request_headers, content=content, dispatcher=dispatcher, **kwargs
            )

        return response

    async def get_authorization_header(self) -> str:
        """Return the ``Authorization`` header value (``Bearer <token>``)."""
        token = await self._get_access_token()
        return f"Bearer {token}"

    async def get_jwt(self) -> DecodedJWT:
        """Return the decoded (unverified) access-token JWT."""
        token = await self._get_access_token()
        return decode_jwt(token)

    async def get_token_response(self) -> AccessTokenResponse:
        """Return the full token response (token + expiry + scopes), using the cache when valid."""
        cached = get_cached_oauth_token(self._cache_key, self._config.scopes or [])
        if cached is not None:
            get_logger("auth.oauth").debug("[OAuthStrategy] Reusing cached access token")
            return cached
        return await self._refresh_token_for_scopes(self._config.scopes)

    def invalidate_token(self) -> None:
        """Invalidate every cached token for this client/method/AM-host identity."""
        invalidate_cached_tokens_for_identity(self._identity_prefix)

    def with_additional_scopes(self, additional_scopes: list[str]) -> OAuthStrategy:
        """Return a new strategy with ``additional_scopes`` merged into the configured scopes."""
        merged = list(dict.fromkeys([*(self._config.scopes or []), *additional_scopes]))
        return OAuthStrategy(
            OAuthConfig(
                client_id=self._config.client_id,
                client_secret=self._config.client_secret,
                scopes=merged,
                account_manager_host=self._config.account_manager_host,
            )
        )

    async def get_access_token_for_cascade(self, candidates: list[list[str]]) -> str:
        """Resolve a scope cascade, returning the first token AM accepts.

        Each candidate is merged with this strategy's base scopes. Pass 1 scans the
        cache for a token satisfying any candidate; pass 2 requests each candidate
        from AM in order, skipping ``invalid_scope`` rejections and rethrowing anything else.
        """
        logger = get_logger("auth.oauth")
        base_scopes = self._config.scopes or []

        # Pass 1: cache scan.
        for candidate in candidates:
            required = list(dict.fromkeys([*base_scopes, *candidate]))
            cached = find_cached_token_satisfying(self._identity_prefix, required)
            if cached is not None:
                logger.debug("[OAuthStrategy] Cache hit: cached token satisfies cascade candidate %s", candidate)
                return cached.access_token

        # Pass 2: try each candidate against AM in order.
        last_error: BaseException | None = None
        for candidate in candidates:
            merged = list(dict.fromkeys([*base_scopes, *candidate]))
            try:
                logger.debug("[OAuthStrategy] Cascade trying scopes %s", candidate)
                token_response = await self._refresh_token_for_scopes(merged)
                return token_response.access_token
            except Exception as error:  # noqa: BLE001 - invalid_scope is expected; others rethrow
                if "invalid_scope" in str(error):
                    logger.debug(
                        "[OAuthStrategy] Cascade candidate %s rejected (invalid_scope), trying next", candidate
                    )
                    last_error = error
                    continue
                raise

        if last_error is not None:
            raise last_error
        raise RuntimeError("All scope cascade candidates failed")

    async def _get_access_token(self) -> str:
        cached = get_cached_oauth_token(self._cache_key, self._config.scopes or [])
        if cached is not None:
            get_logger("auth.oauth").debug("[OAuthStrategy] Reusing cached access token")
            return cached.access_token
        token_response = await self._refresh_token_for_scopes(self._config.scopes)
        return token_response.access_token

    async def _refresh_token_for_scopes(self, scopes: list[str] | None) -> AccessTokenResponse:
        """Fetch a token for a specific scope set, coalescing concurrent callers (single-flight)."""
        cache_key = get_oauth_cache_key(
            self._config.client_id, "client-credentials", self._account_manager_host, scopes
        )
        existing = _PENDING_TOKEN_REQUESTS.get(cache_key)
        if existing is not None:
            get_logger("auth.oauth").debug("[OAuthStrategy] Joining in-flight token request")
            return await existing

        async def _do() -> AccessTokenResponse:
            get_logger("auth.oauth").debug("[OAuthStrategy] Requesting new access token")
            token_response = await self._client_credentials_grant(scopes)
            set_cached_oauth_token(cache_key, token_response)
            return token_response

        task: asyncio.Future[AccessTokenResponse] = asyncio.ensure_future(_do())
        _PENDING_TOKEN_REQUESTS[cache_key] = task
        try:
            return await task
        finally:
            _PENDING_TOKEN_REQUESTS.pop(cache_key, None)

    async def _client_credentials_grant(self, scope_override: list[str] | None = None) -> AccessTokenResponse:
        """Perform the client-credentials grant against Account Manager."""
        logger = get_logger("auth.oauth")
        requested_scopes = scope_override if scope_override is not None else self._config.scopes
        url = f"https://{self._account_manager_host}/dwsso/oauth2/access_token"

        params: dict[str, str] = {"grant_type": "client_credentials"}
        if requested_scopes:
            params["scope"] = " ".join(requested_scopes)
        body = urlencode(params)

        credentials = encode_basic_client_credentials(self._config.client_id, self._config.client_secret)
        headers = {
            "Authorization": f"Basic {credentials}",
            "Content-Type": "application/x-www-form-urlencoded",
        }

        middleware = global_auth_middleware_registry.get_middleware()

        logger.debug("[Auth] Using OAuth client_credentials grant for client: %s", self._config.client_id)
        logger.debug("[Auth REQ] POST %s", url)

        async with httpx.AsyncClient() as client:
            request = client.build_request("POST", url, content=body, headers=headers)
            request = await apply_auth_request_middleware(request, middleware)
            try:
                response = await client.send(request)
            except Exception as err:  # noqa: BLE001 - wrapped only if network-level
                host = httpx.URL(url).host
                raise wrap_network_error(err, operation="OAuth token request", host=host) from err
            response = await apply_auth_response_middleware(request, response, middleware)

            logger.debug("[Auth RESP] POST %s %s", url, response.status_code)

            if not response.is_success:
                error_text = response.text
                raise RuntimeError(
                    f"Failed to get access token: {response.status_code} {response.reason_phrase} - {error_text}"
                )
            data = response.json()

        jwt = decode_jwt(data["access_token"])
        logger.debug("[Auth] JWT payload sub=%s", jwt.payload.get("sub"))

        now = datetime.now(tz=timezone.utc)
        expiration = now + timedelta(seconds=data["expires_in"])
        # AM normally echoes granted scopes; some configs omit `scope`. Fall back to
        # what we requested so cache satisfies-checks (cascade resolution) still work.
        scope = data.get("scope")
        scopes = scope.split(" ") if scope else (requested_scopes or [])

        return AccessTokenResponse(access_token=data["access_token"], expires=expiration, scopes=scopes)
